docs/fsharp/style-guide/formatting.md

```diff
@@ -0,0 +1,517 @@
+
+```
+
+Pattern matching of anonymous functions, starting by `function`, should generally not indent too far. For example, indenting one scope as follows is fine:
+
+```fsharp
+lambdaList
+|> List.map (function
+    | Abs(x, body) -> 1 + sizeLambda 0 body
+    | App(lam1, lam2) -> sizeLambda (sizeLambda 0 lam1) lam2
+    | Var v -> 1)
+```
+
+Pattern matching in functions defined by `let` or `let rec` should be indented 4 spaces after starting of `let`, even if `function` keyword is used:
+
+```fsharp
+let rec sizeLambda acc = function
+    | Abs(x, body) -> sizeLambda (succ acc) body
+    | App(lam1, lam2) -> sizeLambda (sizeLambda acc lam1) lam2
+    | Var v -> succ acc
+```
+
+We do not recommend aligning arrows.
+
+## Formatting try/with expressions
+
+Pattern matching on the exception type should be indented at the same level as `with`.
+
+```fsharp
+try
+    if System.DateTime.Now.Second % 3 = 0 then
+        raise (new System.Exception())
+    else
+        raise (new System.ApplicationException())
+with
+| :? System.ApplicationException ->
+    printfn "A second that was not a multiple of 3"
+| _ ->
+    printfn "A second that was a multiple of 3"
+```
+
+## Formatting function parameter application
+
+In general, most function parameter application is done on the same line.
+
+If you wish to apply parameters to a function on a new line, indent them by one scope.
+
+```fsharp
+// OK
+sprintf "\t%s - %i\n\r"
+     x.IngredientName x.Quantity
+
+// OK
+sprintf
+     "\t%s - %i\n\r"
+     x.IngredientName x.Quantity
+
+// OK
+let printVolumes x =
+    printf "Volume in liters = %f, in us pints = %f, in imperial = %f"
+        (convertVolumeToLiter x)
+        (convertVolumeUSPint x)
+        (convertVolumeImperialPint x)
+```
+
+Anonymous function arguments can be either on next line or with a dangling `fun` on the argument line:
+
+```fsharp
+// OK
+let printListWithOffset a list1 =
+    List.iter (fun elem ->
+        printfn "%d" (a + elem)) list1
+
+// OK, but prefer previous
+let printListWithOffset a list1 =
+    List.iter (
+        fun elem ->
+            printfn "%d" (a + elem)) list1
+```
+
+### Formatting infix operators
+
+Separate operators by spaces. Obvious exceptions to this rule are the `!` and `.` operators.
+
+Infix expressions are OK to lineup on same column:
+
+```fsharp
+acc +
+(sprintf "\t%s - %i\n\r"
+     x.IngredientName x.Quantity)
+
+let function1 arg1 arg2 arg3 arg4 =
+    arg1 + arg2 +
+    arg3 + arg4
+```
+
+### Formatting pipeline operators
+
+Pipeline `|>` should go at the start of a line immediately under the expression being operated on:
+
+```fsharp
+// OK
+let methods2 = System.AppDomain.CurrentDomain.GetAssemblies()
+               |> List.ofArray
+               |> List.map (fun assm -> assm.GetTypes())
+               |> Array.concat
+               |> List.ofArray
+               |> List.map (fun t -> t.GetMethods())
+               |> Array.concat
+
+// OK, but prefer previous
+let methods2 =
```

Comment:
re this:
```
// OK
let methods2 = System.AppDomain.CurrentDomain.GetAssemblies()
               |> List.ofArray
               |> List.map (fun assm -> assm.GetTypes())
               |> Array.concat
               |> List.ofArray
               |> List.map (fun t -> t.GetMethods())
               |> Array.concat
```
This first version is not a good convention in practice and I'd always remove it from the compiler code base. In particular code breaks under rename-refactor.  (I think all the code formatting guidelines should be measured with this in mind).

I haven't read the history of the discussion, so I don't know we're you're at on this.  But I'd recommend putting the second version first, and switching the " OK, but prefer previous"

Reply:
+alot I believe I commented to that effect

My grounds for preferring indenting by one unit on next line are that unindenting when the hard-won indentation is not a multiple of your preferred indentation level often  breaks compilation / varies the meaning in various contexts
(a not unimportant secondary justification is that the further code is indented, the harder it is to review in various contexts.)

Reply:
I'll get a PR rolling

Reply:
I'm going to just remove that first example. I'll want to do a deeper pass in this article as it's mostly just a transcription of the Fantomas doc.